### PR TITLE
[ARCTIC-331]Arctic does not set the primary key field to Required when upgrading the Hive table

### LINKED ISSUE #331: [Bug]: Arctic does not set the primary key field to Required when upgrading the Hive table
### What happened?

Arctic does not set the primary key field to Required when upgrading the Hive table

### Affects Versions

0.3.1

### What engines are you seeing the problem on?

_No response_

### How to reproduce

When the keyed table in the Arctic Hive upgrade table is written by the Flink task, the keyed table cannot be written properly

### Relevant log output

_No response_

### Anything else

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by shendanfengg**
## Why are the changes needed?

*When upgrading the Hive table,  must set isOptional to false for the primary key field

Resolve #331.
## Brief change log

  - *Added a function to convert Hive table schema into Iceberg Schema

## How was this patch tested?
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: core/src/main/java/com/netease/arctic/catalog/BaseArcticCatalog.java [hunk 1]
```diff
@@ -425,6 +427,15 @@ public Transaction newCreateTableTransaction() {
     }
 
     protected void doCreateCheck() {
+      if (primaryKeySpec.primaryKeyExisted()) {
+        Set<String> pkSet = new HashSet<>(primaryKeySpec.fieldNames());
+        schema.columns().forEach(nf -> {
+          if (pkSet.contains(nf.name()) && nf.isOptional()) {
+            throw new IllegalStateException("please check your schema, the primary key nested field must be required " +
```
Comment (zhoujinsong) [inline core/src/main/java/com/netease/arctic/catalog/BaseArcticCatalog.java:434]: `IllegalArgumentException` would be more suitable？

File: core/src/main/java/com/netease/arctic/catalog/BaseArcticCatalog.java [hunk 2]
```diff
@@ -425,6 +427,15 @@ public Transaction newCreateTableTransaction() {
     }
 
     protected void doCreateCheck() {
+      if (primaryKeySpec.primaryKeyExisted()) {
+        Set<String> pkSet = new HashSet<>(primaryKeySpec.fieldNames());
```
Comment (zhoujinsong) [inline core/src/main/java/com/netease/arctic/catalog/BaseArcticCatalog.java:431]: Dose `primaryKeySpec.fieldNames().forEach(column->schema.findField(column).isOptional)` would be more simple?

File: hive/src/main/java/com/netease/arctic/hive/utils/HiveSchemaUtil.java [hunk 3]
```diff
@@ -68,4 +73,28 @@ public static List<FieldSchema> hiveTableFields(Schema schema, PartitionSpec spe
...
+   * @return An Iceberg schema
+   */
+  public static Schema convertHiveSchemaToIcebergSchema(Table hiveTable, List<String> primaryKeys) {
+    List<FieldSchema> hiveSchema = hiveTable.getSd().getCols();
+    hiveSchema.addAll(hiveTable.getPartitionKeys());
+    Set<String> pkSet = new HashSet<>(primaryKeys);
+    Schema schema = org.apache.iceberg.hive.HiveSchemaUtil.convert(hiveSchema);
+    List<Types.NestedField> columns = schema.columns();
+    List<Types.NestedField> columnsWithPk = new ArrayList<>();
+    columns.forEach(nf -> {
```
Comment (zhoujinsong) [inline hive/src/main/java/com/netease/arctic/hive/utils/HiveSchemaUtil.java:91]: Recommend to not use short name `nf`.

**Review by zhoujinsong — COMMENTED** [review]
@shendanfengg `HiveMetaSynchronizer.syncHiveSchemaToArctic` may need to adapt to this change?

File: core/src/main/java/com/netease/arctic/catalog/BaseArcticCatalog.java (hunk 1 above)
Comment (shendanfengg) [reply core/src/main/java/com/netease/arctic/catalog/BaseArcticCatalog.java]: Has been modified.

File: core/src/main/java/com/netease/arctic/catalog/BaseArcticCatalog.java (hunk 2 above)
Comment (shendanfengg) [reply core/src/main/java/com/netease/arctic/catalog/BaseArcticCatalog.java]: Has been modified.

File: hive/src/main/java/com/netease/arctic/hive/utils/HiveSchemaUtil.java (hunk 3 above)
Comment (shendanfengg) [reply hive/src/main/java/com/netease/arctic/hive/utils/HiveSchemaUtil.java]: Has been modified.

File: hive/src/main/java/com/netease/arctic/hive/utils/HiveMetaSynchronizer.java [hunk 4]
```diff
@@ -74,7 +74,9 @@ public static void syncHiveSchemaToArctic(ArcticTable table, HMSClient hiveClien
     try {
       Table hiveTable = hiveClient.run(client -> client.getTable(table.id().getDatabase(), table.id().getTableName()));
       List<FieldSchema> fieldSchemas =  hiveTable.getSd().getCols();
```
Comment (zhoujinsong) [inline hive/src/main/java/com/netease/arctic/hive/utils/HiveMetaSynchronizer.java]: It seems `fieldSchemas ` is no longer needed.

**Review by zhoujinsong — CHANGES_REQUESTED** [review]

**Review by zhoujinsong — APPROVED** [review]
LGTM.

**Pull request merged**